File: packages/hushspec/src/adapters/openai.ts
import type { EvaluationAction, EvaluationResult } from '../evaluate.js';
import type { HushGuard } from '../middleware.js';

export function mapOpenAIToolCall(
  functionName: string,
  functionArgs: string | Record<string, unknown>,
): EvaluationAction {
  const args =
    typeof functionArgs === 'string'
      ? (JSON.parse(functionArgs) as Record<string, unknown>)
      : functionArgs;

  return {
    type: 'tool_call',
    target: functionName,
    args_size:
      typeof functionArgs === 'string'
        ? functionArgs.length
        : JSON.stringify(args).length,
  };
}

export function createOpenAIGuard(
  guard: HushGuard,
): (functionName: string, functionArgs: string | Record<string, unknown>) => EvaluationResult {
  return (
    functionName: string,
    functionArgs: string | Record<string, unknown>,
  ): EvaluationResult => {
    const action = mapOpenAIToolCall(functionName, functionArgs);
    return guard.evaluate(action);
  };
}

File: packages/python/hushspec/adapters/openai.py
from __future__ import annotations

import json

from hushspec.evaluate import EvaluationAction, EvaluationResult
from hushspec.middleware import HushGuard


def map_openai_tool_call(
    function_name: str,
    function_args: str | dict,
) -> EvaluationAction:
    if isinstance(function_args, str):
        args_size = len(function_args)
        json.loads(function_args)
    else:
        args_size = len(json.dumps(function_args))

    return EvaluationAction(
        type="tool_call",
        target=function_name,
        args_size=args_size,
    )


def create_openai_guard(guard: HushGuard):

    def handler(
        function_name: str,
        function_args: str | dict,
    ) -> EvaluationResult:
        action = map_openai_tool_call(function_name, function_args)
        return guard.evaluate(action)

    return handler
